A77-xds-rate-limiting-rlqs.md

```diff
@@ -80,113 +83,383 @@ config:
+   introduce complexities in efficient cache sharding for certain programming
+   languages.
+2. Incrementing `num_request_allowed`/`num_request_denied` bucket counters.
+
+Each gRPC implementation needs to consider what synchronization primitives are
+available in their language to minimize the thread lock time.
+
+#### RLQS Client
+
+The RLQS Client is responsible for communication with the Rate Limit Quota
+Service (RLQS) server. It manages a gRPC stream to the RLQS server, used for
+sending periodic bucket usage reports and receiving new rate limit quota
+assignments. The client handles reconnects and manages the lifecycle of the
+stream.
+
+RLQS Client object will include the following data members:
+
+-   gRPC Channel: From filter config, initialized at instantiation, constant. A
+    channel to the RLQS server.
+-   gRPC Stream: Initialized at instantiation, replaced any time the stream
+    fails (with backoff). Accessed when we get the first data plane RPC for a
+    given bucket and when a report timer fires. A bidirectional gRPC stream to
+    the RLQS server.
+-   Buckets Update Callback: Provided by RLQS Filter state at instantiation,
+    constant, accessed on responses from the RLQS server. A callback to notify
+    the RLQS Filter State of updates to quota assignments.
+
+### Handling Events
+
+#### On LDS/RDS Updates
+
+When receiving an LDS/RDS update, the RLQS filter will perform the following 
+steps for each route:
+
+1.  Parse the new filter config and per-route overrides into an internal RLQS
+    Filter Config object.
+2.  Retrieve the corresponding RLQS Filter State from the RLQS Filter State
+    Cache, creating it if it doesn't exist.
+3.  Generate new data plane RPC handler using the retrieved RLQS Filter State.
+
+Once all routes are parsed, the RLQS filter will release from the RLQS Filter
+State Cache all RLQS Filter State instances no longer referenced by any data
+plane RPC handlers.
 
-The filter parses the config, combines LDS filter config with RDS overrides, and
-generates the `onClientCall` handlers (aka interceptors in Java and Go, and
-filters in C++).
+#### On Data Plane RPC
 
-##### RLQS Cache
+When processing each data plane RPC, the RLQS Filter will ask the RLQS Filter
+State for a rate-limit decision for the RPC. The RLQS Filter State uses the
+matcher tree from the filter config to determine which bucket to use for the
+RPC. It then looks for that RLQS Bucket in the RLQS Bucket Map, creating it if
+it doesn't already exist.
 
-RLQS Cache persists across LDS/RDS updates. It maps unique filter configs to
-RLQS Filter State instances, and provides the thread safety for creating and
-accessing them. Each unique filter config generates a unique RLQS Filter state,
-a 1:1 mapping.
+If a new bucket has been created, the RPC is rate-limited according to bucket's
+default configuration. Then the filter schedules a report on the RLQS stream
```

Comment:
We need to document what fields we will populate in the RLQS requests and what their values will be.

Reply:
See https://github.com/sergiitk/grfc-proposal/blob/a77-rlqs/A77-xds-rate-limiting-rlqs.md#on-sending-usage-reports